### PR TITLE
feat: Separate update event into the update, soft remove and recover events

### PR DESCRIPTION
**Pull request by spotykatch**
Closes: #8398

BREAKING CHANGE: update listeners and subscriber no longer triggered by soft-remove and recover

Fixes: #6179

### Description of change

Separate update event into the update, soft remove and recover events

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change [ Need help with Chinese docs ]
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by spotykatch** [comment]
Don't really get how listeners are registered. I get this output.
```console
BEFORE_UPDATE_LISTENER_CALL
BEFORE_UPDATE_SUBSCRIBER_CALL
AFTER_UPDATE_LISTENER_CALL
AFTER_UPDATE_SUBSCRIBER_CALL
BEFORE_SOFT_REMOVE_SUBSCRIBER_CALL
AFTER_SOFT_REMOVE_SUBSCRIBER_CALL
BEFORE_RECOVER_SUBSCRIBER_CALL
AFTER_RECOVER_SUBSCRIBER_CALL
```

**Comment by spotykatch** [comment]
`callListethis.expressionMap.callListenersners` is false for some reason oO

**Comment by spotykatch** [comment]
LOL. Did not expect checks will be passed so soon) Maybe there should be more checks?
I'll push docs changes. And I think I will need some help here because my English is not good enough to write docs I guess.

**Comment by spotykatch** [comment]
I'll make rebase after review.

**Comment by spotykatch** [comment]
In Chinese documentation, I only changed English words :face_in_clouds:

**Comment by spotykatch** [comment]
Should there be tests for relations?

File: docs/decorator-reference.md [hunk 1]
```diff
@@ -700,6 +706,82 @@ export class Post {
 
 Learn more about [listeners](listeners-and-subscribers.md).
 
+#### `@BeforeSoftRemove`
+
+You can define a method with any name in the entity and mark it with `@BeforeRemove`
```
Comment (pleerock) [inline docs/decorator-reference.md:711]: I think `@BeforeRemove` should be `@BeforeSoftRemove` here

File: docs/decorator-reference.md [hunk 2]
```diff
@@ -700,6 +706,82 @@ export class Post {
...
+and TypeORM will call it before a entity is soft removed using repository/manager `remove`.
+Example:
+
+```typescript
+@Entity()
+export class Post {
+
+    @BeforeRemove()
```
Comment (pleerock) [inline docs/decorator-reference.md:719]: same issue

File: docs/decorator-reference.md [hunk 3]
```diff
@@ -700,6 +706,82 @@ export class Post {
...
+        this.status = "soft-removed";
+    }
+}
+```
+
+Learn more about [listeners](listeners-and-subscribers.md).
+
+#### `@AfterSoftRemove`
+
+You can define a method with any name in the entity and mark it with `@AfterRemove`
```
Comment (pleerock) [inline docs/decorator-reference.md:730]: same issue

File: docs/decorator-reference.md [hunk 4]
```diff
@@ -700,6 +706,82 @@ export class Post {
...
+and TypeORM will call it after the entity is soft removed using repository/manager `remove`.
+Example:
+
+```typescript
+@Entity()
+export class Post {
+
+    @AfterRemove()
```
Comment (pleerock) [inline docs/decorator-reference.md:738]: same issue

File: docs/decorator-reference.md [hunk 5]
```diff
@@ -700,6 +706,82 @@ export class Post {
...
+
+You can define a method with any name in the entity and mark it with `@BeforeRecover`
+and TypeORM will call it before a entity is recovered using repository/manager `recover`.
+Example:
+
+```typescript
+@Entity()
+export class Post {
+
+    @BeforeRemove()
```
Comment (pleerock) [inline docs/decorator-reference.md:757]: same issue

File: docs/decorator-reference.md [hunk 6]
```diff
@@ -700,6 +706,82 @@ export class Post {
...
+        this.status = "recovered";
+    }
+}
+```
+
+Learn more about [listeners](listeners-and-subscribers.md).
+
+#### `@AfterRecover`
+
+You can define a method with any name in the entity and mark it with `@AfterRemove`
```
Comment (pleerock) [inline docs/decorator-reference.md:768]: same issue

File: docs/decorator-reference.md [hunk 7]
```diff
@@ -700,6 +706,82 @@ export class Post {
...
+and TypeORM will call it after the entity is recovered using repository/manager `recover`.
+Example:
+
+```typescript
+@Entity()
+export class Post {
+
+    @AfterRemove()
```
Comment (pleerock) [inline docs/decorator-reference.md:776]: same issue

File: docs/listeners-and-subscribers.md [hunk 8]
```diff
@@ -8,6 +8,10 @@
     * [`@AfterUpdate`](#afterupdate)
     * [`@BeforeRemove`](#beforeremove)
     * [`@AfterRemove`](#afterremove)
+    * [`@BeforeSoftRemove`](#beforeremove)
+    * [`@AfterSoftRemove`](#afterremove)
+    * [`@BeforeRecover`](#beforeremove)
+    * [`@AfterRecover`](#afterremove)
```
Comment (pleerock) [inline docs/listeners-and-subscribers.md:14]: incorrect links all above ^

File: docs/listeners-and-subscribers.md [hunk 9]
```diff
@@ -89,6 +94,7 @@ export class Post {
 
 You can define a method with any name in the entity and mark it with `@AfterUpdate`
 and TypeORM will call it after an existing entity is updated using repository/manager `save`.
+> :warning: **@AfterUpdate is no longer triggered by `softRemove` and `recover`
```
Comment (pleerock) [inline docs/listeners-and-subscribers.md:97]: I don't think we need these warnings above, just remove the all

File: docs/listeners-and-subscribers.md [hunk 10]
```diff
@@ -136,6 +142,74 @@ export class Post {
...
+
+You can define a method with any name in the entity and mark it with `@BeforeSoftRemove`
+and TypeORM will call it before a entity is soft removed using repository/manager `softRemove`.
+Example:
+
+```typescript
+@Entity()
+export class Post {
+
+    @BeforeRemove()
```
Comment (pleerock) [inline docs/listeners-and-subscribers.md:155]: same issue

File: docs/listeners-and-subscribers.md [hunk 11]
```diff
@@ -136,6 +142,74 @@ export class Post {
...
+
+You can define a method with any name in the entity and mark it with `@BeforeRecover`
+and TypeORM will call it before a entity is recovered using repository/manager `recover`.
+Example:
+
+```typescript
+@Entity()
+export class Post {
+
+    @BeforeRemove()
```
Comment (pleerock) [inline docs/listeners-and-subscribers.md:189]: same issue

File: docs/decorator-reference.md [hunk 12]
```diff
@@ -700,6 +706,82 @@ export class Post {
...
+    }
+}
+```
+
+Learn more about [listeners](listeners-and-subscribers.md).
+
+#### `@AfterSoftRemove`
+
+You can define a method with any name in the entity and mark it with `@AfterRemove`
+and TypeORM will call it after the entity is soft removed using repository/manager `remove`.
```
Comment (pleerock) [inline docs/decorator-reference.md:731]: using `softRemove` ?

File: docs/decorator-reference.md [hunk 13]
```diff
@@ -700,6 +706,82 @@ export class Post {
 
 Learn more about [listeners](listeners-and-subscribers.md).
 
+#### `@BeforeSoftRemove`
+
+You can define a method with any name in the entity and mark it with `@BeforeRemove`
+and TypeORM will call it before a entity is soft removed using repository/manager `remove`.
```
Comment (pleerock) [inline docs/decorator-reference.md:712]: using `softRemove` ?

File: docs/zh_CN/decorator-reference.md [hunk 14]
```diff
@@ -28,6 +28,10 @@
     * [`@AfterUpdate`](#afterupdate)
     * [`@BeforeRemove`](#beforeremove)
     * [`@AfterRemove`](#afterremove)
+    * [`@BeforeSoftRemove`](#beforeremove)
+    * [`@AfterSoftRemove`](#afterremove)
+    * [`@BeforeRecover`](#beforeremove)
+    * [`@AfterRecover`](#afterremove)
```
Comment (pleerock) [inline docs/zh_CN/decorator-reference.md:34]: links are broken ^

File: docs/zh_CN/listeners-and-subscribers.md [hunk 15]
```diff
@@ -7,6 +7,10 @@
   * [`@AfterUpdate`](#afterupdate)
   * [`@BeforeRemove`](#beforeremove)
   * [`@AfterRemove`](#afterremove)
+  * [`@BeforeSoftRemove`](#beforeremove)
+  * [`@AfterSoftRemove`](#afterremove)
+  * [`@BeforeRecover`](#beforeremove)
+  * [`@AfterRecover`](#afterremove)
```
Comment (pleerock) [inline docs/zh_CN/listeners-and-subscribers.md:13]: links are broken ^

File: docs/zh_CN/listeners-and-subscribers.md [hunk 16]
```diff
@@ -119,6 +123,66 @@ export class Post {
...
+  @BeforeSoftRemove()
+  updateStatus() {
+    this.status = "soft-removed";
+  }
+}
+```
+
+### `@AfterSoftRemove`
+
+你可以在实体中定义一个具有任何名称的方法，并使用`@ AfterSoftRemove`标记它，TypeORM 将在使用 repository/manager `softRemove`删除实体后调用它。
```
Comment (pleerock) [inline docs/zh_CN/listeners-and-subscribers.md:143]: redundant space

Comment (pleerock) [reply docs/zh_CN/listeners-and-subscribers.md:143]: and missing spaces

File: docs/zh_CN/listeners-and-subscribers.md [hunk 17]
```diff
@@ -119,6 +123,66 @@ export class Post {
...
+  @AfterSoftRemove()
+  updateStatus() {
+    this.status = "soft-removed";
+  }
+}
+```
+
+### `@BeforeRecover`
+
+你可以在实体中定义具有任何名称的方法，并使用`@BeforeRecover`标记它，并且 TypeORM 将在使用 repository/manager `recover`删除实体之前调用它。
```
Comment (pleerock) [inline docs/zh_CN/listeners-and-subscribers.md:158]: and missing spaces

File: docs/zh_CN/listeners-and-subscribers.md [hunk 18]
```diff
@@ -119,6 +123,66 @@ export class Post {
...
+  @BeforeRecover()
+  updateStatus() {
+    this.status = "recovered";
+  }
+}
+```
+
+### `@AfterRecover`
+
+你可以在实体中定义一个具有任何名称的方法，并使用`@ AfterRecover`标记它，TypeORM 将在使用 repository/manager `recover`删除实体后调用它。
```
Comment (pleerock) [inline docs/zh_CN/listeners-and-subscribers.md:173]: and missing spaces

File: src/query-builder/SoftDeleteQueryBuilder.ts [hunk 19]
```diff
@@ -63,9 +63,12 @@ export class SoftDeleteQueryBuilder<Entity> extends QueryBuilder<Entity> impleme
...
             }
 
-            // call before updation methods in listeners and subscribers
+            // call before soft remove and recover methods in listeners and subscribers
             if (this.expressionMap.callListeners === true && this.expressionMap.mainAlias!.hasMetadata) {
-                await queryRunner.broadcaster.broadcast("BeforeUpdate", this.expressionMap.mainAlias!.metadata);
+                if (this.expressionMap.queryType === "soft-delete")
+                    await queryRunner.broadcaster.broadcast("BeforeSoftRemove", this.expressionMap.mainAlias!.metadata);
+                else if (this.expressionMap.queryType === "restore")
+                    await queryRunner.broadcaster.broadcast("AfterSoftRemove", this.expressionMap.mainAlias!.metadata);
```
Comment (pleerock) [inline src/query-builder/SoftDeleteQueryBuilder.ts:71]: I guess it should be `BeforeRecover` here?

**Comment by pleerock** [comment]
Since this PR contains many small typos and copy-paste issues I recommend to go through the code once again (in case if I missing something) and make sure no issues left. Good job on a PR itself

**Comment by spotykatch** [comment]
Oh, sorry, I guess those two lines are so bad. I'll make changes in 2-3 days. Don't have time right now.
```ts
export interface SoftRemoveEvent<Entity> extends RemoveEvent<Entity> {}
```
```ts
export interface RecoverEvent<Entity> extends RemoveEvent<Entity> {}
```
Should have `databaseEntity`.

**Comment by pleerock** [comment]
@spotykatch were you able to fix the remaining issues before we can merge this one?

**Comment by spotykatch** [comment]
Chinese docs are the only got left. I hope Chinese users will fix it themselves 😺

**Comment by pleerock** [comment]
image

**Pull request merged**